File: 0040-combination-sum-ii/0040-combination-sum-ii.py
class Solution(object):
    def combinationSum2(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        def backtrack_comb(start, arr, target) :
            if target == 0 :
                result.append(arr[:])
                return
            prev = -1
            for i in range(start, len(candidates)) :
                if i > start and candidates[i] == prev :
                    continue
                if candidates[i] > target: break
                prev = candidates[i]
                backtrack_comb(i+1, arr+[candidates[i]], target-candidates[i])

        result = []
        candidates.sort()
        backtrack_comb(0, [], target)
        return result
  

File: 0040-combination-sum-ii/0040-combination-sum-ii.cpp
class Solution {
public:
    void backtrackComb(int start, int target, vector<int>& candidates, vector<int>& arr, vector<vector<int>>& result){
        if (target == 0 ){
            result.push_back(arr);
            return;
        }
        int prev = -1;
        for (int idx=start; idx < candidates.size(); idx++){
            if (candidates[idx] == prev) continue;
            if (candidates[idx] <= target) {
                prev = candidates[idx];
                arr.push_back(candidates[idx]);
                backtrackComb(idx+1, target-candidates[idx], candidates, arr, result);
                arr.pop_back();
            }
        }

    }
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<vector<int>> result;
        vector<int> arr;
        // sort candidates
        sort(candidates.begin(), candidates.end(), [](int a, int b){
            return a < b;
        }
        );
        backtrackComb(0, target, candidates, arr, result);
        return result;        
    }
};
